Draft4 support

andreineculau opened the pull request:
Port of https://github.com/klarna/jesse/pull/36 by @emfa

---

Adds support for the new keywords, removes deprecated ones and changes the behavior of those that has been changed.

gadgetci commented on src/jesse_validator_draft4.erl:
```diff
@@ -0,0 +1,848 @@
+    case jesse_lib:is_json_object(Schema) of
+      true ->
+        State1 = jesse_state:set_current_schema(Schema, State0),
+        State2 = jesse_schema_validator:validate_with_state(Value, Schema, State1),
```
According to **Elvis**:
> Line 736 is too long:         State2 = jesse_schema_validator:validate_with_state(Value, Schema, State1),.

andreineculau commented:
Port of https://github.com/klarna/jesse/issues/16 by @zuiderkwast

---

# What needs to be done

* Small adjustments of current code (some renamed things: divisibleBy to multipleOf, readonly to readOnly, ...)
* Add the new things that are not yet implemented to the README.

# Considerations

* Should we keep backwards compatibility with draft 3 when possible (keep renamed and/or deleted things: readonly, divisibleBy, extends, disallow, type=any) or drop them?
* A flag for which draft to validate against? (Intuitively, I doen feel it's an attractive thing to do.)

# Change logs for draft 4

Core schema:

* Initial draft.
* Salvaged from draft v3.
* Mandate the use of JSON Reference, JSON Pointer.
* Define the role of "id". Define URI resolution scope.
* Add interoperability considerations.

Validation schema:

* Initial draft.
* Salvaged from draft v3.
* Redefine the "required" keyword.
* Remove "extends", "disallow"
* Add "anyOf", "allOf", "oneOf", "not", "definitions", "minProperties", "maxProperties".
"dependencies" member values can no longer be single strings; at least one element is required in a property dependency array.
* Rename "divisibleBy" to "multipleOf".
* "type" arrays can no longer have schemas; remove "any" as a possible value.
Rework the "format" section; make support optional.
* "format": remove attributes "phone", "style", "color"; rename "ip-address" to "ipv4"; add references for all attributes.
* Provide algorithms to calculate schema(s) for array/object instances.
* Add interoperability considerations.

Hyper schema:

* Resolution of link URIs ("href") is now affected by rel="self" links on the instance
* Define "title" for LDOs
* Use URI Templates for the "href" property
* Split hyper-schema definition out from main schema.
* Capitalised the T in "encType", and the O in "readOnly"
* Moved "mediaType" and "contentEncoding" to the new "media" property (renamed "type" and "binaryEncoding")
* Added "mediaType" property to LDOs
* Replaced "slash-delimited" fragment resolution with "json- pointer".
* Added "template" LDO attribute.
* Improved wording of sections.

andreineculau commented:
closing in favour of #5
